destination/writer/writer.go

```diff
@@ -0,0 +1,117 @@
+// Copyright © 2022 Meroxa, Inc. & Yalantis
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package writer
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+
+	sdk "github.com/conduitio/conduit-connector-sdk"
+	"go.mongodb.org/mongo-driver/bson"
+	"go.mongodb.org/mongo-driver/mongo"
+)
+
+const (
+	// idFieldName contains default reserved primary key from MongoDB.
+	idFieldName = "_id"
+
+	// setCommand contains command, that used during Update query.
+	setCommand = "$set"
+)
+
+// ErrEmptyKey occurs when a record has an empty key and an operation is update or delete.
+var ErrEmptyKey = errors.New("empty key")
+
+// Writer implements a writer logic for Mongo destination.
+type Writer struct {
+	collection *mongo.Collection
+}
+
+// NewWriter creates new instance of the Writer.
+func NewWriter(collection *mongo.Collection) *Writer {
+	writer := &Writer{
+		collection: collection,
+	}
+
+	return writer
+}
+
+// Write writes a sdk.Record into a Destination.
+func (w *Writer) Write(ctx context.Context, record sdk.Record) error {
+	if err := sdk.Util.Destination.Route(ctx, record,
+		w.insert,
+		w.update,
+		w.delete,
+		w.insert,
+	); err != nil {
+		return fmt.Errorf("route %s: %w", record.Operation, err)
+	}
+
+	return nil
+}
+
+func (w *Writer) insert(ctx context.Context, record sdk.Record) error {
+	payload := make(sdk.StructuredData)
+	if err := json.Unmarshal(record.Payload.After.Bytes(), &payload); err != nil {
+		return fmt.Errorf("unmarshal payload: %w", err)
+	}
+
+	if _, err := w.collection.InsertOne(ctx, bson.M(payload)); err != nil {
+		return fmt.Errorf("insert one: %w", err)
+	}
+
+	return nil
```

Comment:
I'm thinking about providing an option to use the record keys if they exist. If someone wants to keep MongoDB in sync with another source, they may want to keep the IDs in sync as well.

We can make that option configurable (i.e. use record keys vs auto-generate ID). If it's a significant amount of work, we can do it in a separate issue (but we should document it).

Reply:
But we use `sdk.Record.Key` for `Update` and `Delete` operations.
Or are we talking about `Insert` operation?

Reply:
Yes, we're talking about the `insert` operation. Now that I'm thinking of it, I think it's good to just create an issue for this. I believe most products will use auto-generated IDs in MongoDB, so being able to provide your own keys/IDs is probably not a big thing right now. WDYT?

Reply:
IIUC, you meant to add possibility to fill `_id` field with parameter value from `sdk.Record.Keys` by a key given in configuration?

E.g.:
There is a `sdk.Record.Key` -`{"filed_1":"uuid_1","filed_2":"uuid_2"}`.
Add the configuration field `keyID` and fill in it with `field_2`.

If the config field `keyID` is not empty - get the value from the `sdk.Record.Key` by this key.
And then set the `_id` field of new record with the value from previous step.

Is that what you meant?

Reply:
@voscob That's one of the possibilities, yes. Another not to use the _id field, but create unique constraints. But as mentioned, I don't think it's needed for this PR.

Reply:
Got it.
I'll create an issue.
Bit I want to say it's not possible to use another unique constraint instead of `_id` field name.
https://www.mongodb.com/docs/manual/reference/glossary/#std-term-_id

Reply:
I've created a [new issue](https://github.com/conduitio-labs/conduit-connector-mongo/issues/40) with the link to this thread.

Reply:
Thank you!